causallearn/utils/cit.py

```diff
@@ -15,78 +15,142 @@
+            assert self.pvalue_cache['method_name'] == method_name, "CI test method name mismatch." # a loaded cache
+            assert self.pvalue_cache['parameters_hash'] == parameters_hash, "CI test method parameters mismatch."
+
+    def assert_input_data_is_valid(self, allow_nan=False, allow_inf=False):
+        assert allow_nan or not np.isnan(self.data).any(), "Input data contains NaN. Please check."
+        assert allow_inf or not np.isinf(self.data).any(), "Input data contains Inf. Please check."
+
+    def save_to_local_cache(self):
+        if not self.cache_path is None and time.time() - self.last_time_cache_saved > self.SAVE_CACHE_CYCLE_SECONDS:
+            with codecs.open(self.cache_path, 'w') as fout: fout.write(json.dumps(self.pvalue_cache, indent=2))
+            self.last_time_cache_saved = time.time()
+
+    def get_formatted_XYZ_and_cachekey(self, X, Y, condition_set):
+        '''
+        reformat the input X, Y and condition_set to
+            1. convert to built-in types for json serialization
+            2. handle multi-dim unconditional variables (for kernel-based)
+            3. basic check for valid input (X, Y no overlap with condition_set)
+            4. generate unique and hashable cache key
+
+        Parameters
+        ----------
+        X: int, or np.*int*
+        Y: int, or np.*int*
+        condition_set: Iterable<int | np.*int*>
+
+        Returns
+        -------
+        Xs: List<int>, sorted. may swapped with Ys for cache key uniqueness.
+        Ys: List<int>, sorted.
+        condition_set: List<int>
+        cache_key: string. Unique for <X,Y|S> in any input type or order.
+        '''
+        def _stringize(ulist1, ulist2, clist):
+            # ulist1, ulist2, clist: list of ints, sorted.
+            _strlst  = lambda lst: '.'.join(map(str, lst))
+            return f'{_strlst(ulist1)};{_strlst(ulist2)}|{_strlst(clist)}' if len(clist) > 0 else \
+                   f'{_strlst(ulist1)};{_strlst(ulist2)}'
+
+        # every time when cit is called, auto save to local cache.
+        self.save_to_local_cache()
+
+        METHODS_SUPPORTING_MULTIDIM_DATA = ["kci"]
+        if condition_set is None: condition_set = []
+        # 'int' to convert np.*int* to built-in int; 'set' to remove duplicates; sorted for hashing
+        condition_set = sorted(set(map(int, condition_set)))
+
+        # usually, X and Y are 1-dimensional index (in constraint-based methods)
+        if self.method not in METHODS_SUPPORTING_MULTIDIM_DATA:
+            X, Y = (int(X), int(Y)) if (X < Y) else (int(Y), int(X))
+            assert X not in condition_set and Y not in condition_set, "X, Y cannot be in condition_set."
+            return [X], [Y], condition_set, _stringize([X], [Y], condition_set)
+
+        # also to support multi-dimensional unconditional X, Y (usually in kernel-based tests)
+        Xs = sorted(set(map(int, X))) if isinstance(X, Iterable) else [int(X)]  # sorted for comparison
```

Comment:
why not force X to be List<int> type?

Personally I always prefer to make the variable typed --- it can remove lots of potential bug and make the code much cleaner.

Reply:
In all of our constraint-based methods, X and Y are assumed to be integers.

If we force X as List type, all the codes related to cit calls in pc, fci, ... will need to be changed.

Overall, integers X and Y should always be the first-class citizen in CITests. Multi-dim X Y only works in KCI, and not in constaint-based methods, but somewhere else (e.g., GIN, you name it).

Reply:
ic, sounds good.